_posts/2016-01-22-monitoring-an-external-tcp-service-in-elixir.md

```diff
@@ -0,0 +1,378 @@
+    host = Keyword.get(opts, :host, "localhost") |> String.to_char_list
+    port = Keyword.fetch!(opts, :port)
+    %State{host: host, port: port}
+  end
+end
+```
+
+We introduce a `State` struct which gets populated from `opts`. We can then adapt the rest of the code to use it. This also simplifies the callbacks code, as we don't have pattern match on the state tuple anymore. We can now more comfortably handle the maximum number of retries feature.
+
+Tracking the failure count can be implemented as follows:
+
+```elixir
+defmodule State do
+  defstruct host: "localhost",
+            port: 1234,
+            failure_count: 0
+end
+
+...
+
+def init(opts) do
+  state = opts_to_initial_state(opts)
+  case :gen_tcp.connect(state.host, state.port, []) do
+    {:ok, _socket} ->
+      {:ok, state}
+    {:error, _reason} ->
+      {:ok, %{state | failure_count: 1}, @retry_interval}
+  end
+end
+
+def handle_info(:timeout, state = %State{failure_count: failure_count}) do
+  case :gen_tcp.connect(state.host, state.port, []) do
+    {:ok, _socket} ->
+      {:noreply, %{state | failure_count: 0}}
+    {:error, _reason} ->
+      {:noreply, %{state | failure_count: failure_count + 1}, @retry_interval}
+  end
+end
+```
+
+We add a new property to `State` and update/reset its value accordingly depending on the outcome of every `:gen_tcp.connect/3` call.
+
+Tracking the failure count is just the first half of this feature: next is stopping the process when reaching 10 consecutive failures.
+
+```elixir
+defmodule HSC do
+  use GenServer
+  @retry_interval 1000
+  @max_retries 10
+
+  ...
+
+  def handle_info(:timeout, state = %State{failure_count: failure_count}) do
+    if failure_count <= @max_retries do
+      case :gen_tcp.connect(state.host, state.port, []) do
+        {:ok, _socket} ->
+          {:noreply, %{state | failure_count: 0}}
+        {:error, _reason} ->
+          {:noreply, %{state | failure_count: failure_count + 1}, @retry_interval}
+      end
+    else
+      {:stop, :max_retry_exceeded, state}
+    end
+  end
+end
+```
+
+We stop the worker by returning `{:stop, reason, state}` as we did in the beginning. At this point the worker will be restarted by the supervisor and will conform to its strategy. This means that we need to be careful with numbers.
+
+By default the `Supervisor` will restart this worker a maximum of 3 times over 5 seconds (see [the documentation for `supervise/2`](http://elixir-lang.org/docs/stable/elixir/Supervisor.Spec.html#supervise/2) for more details on how to change that), while the worker's lifetime, in case of continuous failure, is at least 10 seconds (1 second interval, 10 retries). With this configuration, **it will never crash the top level supervisor**. To do that we would need to tweak intervals and retries to reach a good balance, especially if we consider how long it would take to restore/replace database connectivity (e.g. promoting another server or rerouting traffic).
```

Comment:
If you really want to crash the whole system on unsuccessful reconnect, I think it would be simpler to just call [:init.stop/1](http://erlang.org/doc/man/init.html#stop-1) instead of tweaking restart frequencies to bubble the failure to the top-level supervisor. 

Though in my experience, a more likely choice is to keep the system running with a partial service, and then resume the full service when reconnection eventually happens.